README
NAME
    Parse::Snort - Parse and create Snort rules

VERSION
    Version 0.01

SYNOPSIS
        use Parse::Snort;

        my $rule = Parse::Snort->new(
          action => 'alert',
          proto => 'tcp',
          src => '$HOME_NET', src_port => 'any',
          direction => '->'
          dst =>'$EXTERNAL_NET', dst_port => 'any'
        );

        $rule->action("pass");

        $rule->opts(
            [ 'depth' => 50 ],
            [ 'offset' => 0 ],
            [ 'content' => "perl6" ],
            [ "nocase" ]
        );

        my $rule = Parse::Snort->new();
        $rule->parse('pass tcp $HOME_NET any -> $EXTERNAL_NET 6667;');
        $rule->msg("IRC server");
        my $rule_string = $rule->as_string;
    );

METHODS
    The following methods can be used to read or modify parts of a rule.

    new($rule_string), new($rule_element_ref)
        This function will create a new "Parse::Snort" object. You may pass
        nothing, a string containing a properly formatted Snort rule, or a
        gash reference of rule elements and options.

        $rule_string
              $rule_string = 'alert tcp $EXTERNAL_NET any -> $HOME_NET any (msg:"perl 6 download detected\; may the world rejoice!";depth:150; offset:0; content:"perl-6.0.0"; nocase;)'

        $rule_element_hashref
              $rule_element_hashref = {
                action => 'alert',
                proto => 'tcp',
                src => '$EXTERNAL_NET', src_port => 'any',
                direction => '->',
                dst => '$HOME_NET', dst_port => 'any',
                opts => [
                    [ 'msg' => ':"perl 6 download detected\; may the world rejoice!"' ],
                    [ 'depth' => 150 ],
                    [ 'offset' => 0 ].
                    [ 'content' => 'perl-6.0.0' ],
                    [ 'nocase' ],
                ],
      
              };

    parse($rule_string)
        The parse method can be used to parse a snort rule string after
        new() has been called. The rule object will be populated with the
        parsed version of $rule_string, overwriting any previously defined
        values in the object.

          $rule_object->parse($rule_string);

  METHODS FOR RULE ELEMENTS
    The following methods read or modify the various rule elements.

    action
        The rule action. Generally one of the following: "alert", "pass",
        "drop", "sdrop", or "log".

    proto
        The protocol of the rule. Generally one of the following: "tcp",
        "udp", "ip", or "icmp".

    src The source IP address for the rule. Generally a dotted decimal IP
        address, Snort $HOME_NET variable, or CIDR block notation.

    src_port
        The source port for the rule. Generally a static port, or a
        contigious range of ports.

    direction
        The direction of the rule. One of the following: "-"> "<"> or "<-".

    dst The destination IP address for the rule. Same format as "src"

    dst_port
        The destination port for the rule. Same format as "src"

    opts($opts_array_ref), opts($opts_string)
        The opts method can be used to read existing options of a parsed
        rule, or set them. The method takes two forms of arguments, either
        an Array of Arrays, or a rule string.

        $opts_array_ref
              $opts_array_ref = [
                   [ 'msg' => ':"perl 6 download detected\; may the world rejoice!"' ],
                   [ 'depth' => 150 ],
                   [ 'offset' => 0 ].
                   [ 'content' => 'perl-6.0.0' ],
                   [ 'nocase' ],
              ]

        $opts_string
              $opts_string='(msg:"perl 6 download detected\; may the world rejoice!";depth:150; offset:0; content:"perl-6.0.0"; nocase;)';

            The parenthesis surround the series of "key:value;" pairs are
            optional.

  HELPER METHODS FOR OPTIONS
    sid, rev, msg, classtype
        The "sid", "rev", "msg", and "classtype" methods allow direct access
        to the rule option of the same name

          my $sid = $rule_obj->sid(); # reads the sid of the rule
          $rule_obj->sid($sid); # sets the sid of the rule
          ... etc ...

    references
        The "references" method returns an array reference of the references
        in the rule. Each reference is an array, in [ 'reference_type' =>
        'reference_value' ] format. To modify references, use the "opts"
        method.

    as_string
        The "as_string" method returns a string that matches the normal
        Snort rule form of the object. This is what you want to use to write
        a rule to an output file that will be read by Snort.

AUTHOR
        Richard G Harman Jr, "<perl-cpan at richardharman.com>"

BUGS
        Please report any bugs or feature requests to "bug-parse-snort at
        rt.cpan.org", or through the web interface at
        <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Parse-Snort>. I will
        be notified, and then you'll automatically be notified of progress
        on your bug as I make changes.

SUPPORT
        You can find documentation for this module with the perldoc command.

            perldoc Parse::Snort

        You can also look for information at:

        * AnnoCPAN: Annotated CPAN documentation
            <http://annocpan.org/dist/Parse-Snort>

        * CPAN Ratings
            <http://cpanratings.perl.org/d/Parse-Snort>

        * RT: CPAN's request tracker
            <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Parse-Snort>

        * Search CPAN
            <http://search.cpan.org/dist/Parse-Snort>

ACKNOWLEDGEMENTS
        MagNET #perl for putting up with me :)

COPYRIGHT & LICENSE
        Copyright 2007 Richard Harman, all rights reserved.

        This program is free software; you can redistribute it and/or modify
        it under the same terms as Perl itself.



Changes
Revision history for Parse-Snort

{{$NEXT}}

    * Add support for preprocessed rule parsing
    * Add Parse::Snort::Strict to validate rules
    * Add state() to indicate an active or inactive rule.
      Inactive rules are stringified as a comment.

0.07    May 5 2018

    * Fix testsuite for flow() accessor

0.06    May 20 2011

    * Third time is the charm, fixing a bug with options that would get truncated w/ a "false" value in perl.

0.05    Dec 1 2010
    * Second release, fixing a whitespace bug, and added some tests.

0.01    May 7 2007
    * First version, released on an unsuspecting world.



lib/Parse/Snort/Strict.pm
package Parse::Snort::Strict;
use base qw(Parse::Snort);

use strict;
use warnings;
use Carp qw(croak);
use List::Util qw(any);
use Sub::Util qw(set_subname);

our $VERSION = "0.8";

# valid values for rule parts
my $rule_parts_validation = {
  action => [qw( alert pass drop sdrop log activate dynamic reject )],
  proto => [qw( tcp udp ip icmp )],
  direction => [qw( -> <> <- )],
};

# method generator for simple rule parts, copypasta reduction
{
  my $generator = sub {
    # closures are teh awesome.
    my ($part,$value_ref) = @_;
    my $method = "SUPER::$part";

    return sub {
      my ($self,$value) = @_;

      # do validation
      croak "Invalid rule $part: '$value'" unless (any { $value eq $_ } @{ $value_ref });

      # call parent's method for value setting
      $self->$method($value);
    };
  };

  no strict qw(refs);
  while (my ($part,$value_ref) = each %$rule_parts_validation) {
    *{$part} = set_subname($part,$generator->($part,$value_ref));
  }
  use strict qw(refs);
}

# TODO: validate formatting of src/dst address and port, make sure they look like $VARIABLEs or [a:range]

1;

__END__;

=head1 NAME

Parse::Snort::Strict - Parse Snort rules with validation of the rules

=head1 DESCRIPTION

Parse Snort rules with validation regarding rule action, protocol and direction.  Look at L<Parse::Snort> for more usage detail, as this is a subclass of it.

=head1 SYNOPSIS

    use Parse::Snort::Strict;
    use Try::Tiny;

    my $rule = Parse::Snort::Strict->new();
    try {
        $rule->parse($text);
    }
    catch {
        warn "Unable to parse rule: $_";
    };

=head1 METHODS

=head2 action

You can only have the following actions

=over

=item alert

generate an alert using the selected alert method, and then

=item log

log the packet

=item pass

ignore the packet

=item activate

alert and then turn on another dynamic rule

=item dynamic

remain idle until activated by an activate rule , then act as a log rule

=item drop

block and log the packet

=item reject

block the packet, log it, and then send a TCP reset if the protocol is
TCP or an ICMP port unreachable message if the protocol is UDP.

=item sdrop

block the packet but do not log it.

=back

=head2 proto

You can only have the following protocols:

=over

=item tcp

=item udp

=item ip

=item icmp

=back

=head2 direction

You can Only have the following directions

=over

=item C<< -> >>

=item C<< <> >>

=item C<< <- >>

=back

=cut


lib/Parse/Snort.pm
package Parse::Snort;

use strict;
use warnings;
use base qw(Class::Accessor);
use List::Util qw(first);
use Carp qw(carp);

our $VERSION = '0.9';

=head1 NAME

Parse::Snort - Parse and create Snort rules

=head1 VERSION

Version 0.9

=head1 SYNOPSIS

    use Parse::Snort;

    my $rule = Parse::Snort->new(
      action => 'alert',
      proto => 'tcp',
      src => '$HOME_NET', src_port => 'any',
      direction => '->'
      dst => '$EXTERNAL_NET', dst_port => 'any'
    );

    $rule->action("pass");

    $rule->opts(
      [ 'depth' => 50 ],
      [ 'offset' => 0 ],
      [ 'content' => "perl6" ],
      [ "nocase" ]
    );

    my $rule = Parse::Snort->new();
    $rule->parse('pass tcp $HOME_NET any -> $EXTERNAL_NET 6667;');
    $rule->msg("IRC server");
    my $rule_string = $rule->as_string;
);

=cut

our @RULE_ELEMENTS_REQUIRED = qw/ action proto src src_port direction dst dst_port /;
our @RULE_ELEMENTS = ( @RULE_ELEMENTS_REQUIRED, 'opts' );

# create the accessors for the standard parts (note; opts comes later)
__PACKAGE__->mk_accessors(@RULE_ELEMENTS_REQUIRED);


=head1 METHODS

These are the object methods that can be used to read or modify any part of a Snort rule.  B<Please note: None of these methods provide any sort of input validation to make sure that the rule makes sense, or can be parsed at all by Snort.>  

If input validation is required, check out the L<Parse::Snort::Strict> module.

=head2 new ()

Create a new C<Parse::Snort> object, and return it.  There are a couple of options when creating the object:

=over 4

=item new ( )

Create an unpopulated object, that can be filled in using the individual rule element methods, or can be populated with the L<< parse|Parse::Snort/"PARSE" >> method.

=item new ( $rule_string )

Create an object based on a plain text Snort rule, all on one line.  This module doesn't understand the UNIX style line continuations (a backslash at the end of the line) that Snort does.

  $rule_string = 'alert tcp $EXTERNAL_NET any -> $HOME_NET any (msg:"perl 6 download detected\; may the world rejoice!";depth:150; offset:0; content:"perl-6.0.0"; nocase;)'


=item new ( $rule_element_hashref )

Create an object baesd on a prepared hash reference similar to the internal strucutre of the L<Parse::Snort> object.

    $rule_element_hashref = {
        action    => 'alert',
        proto     => 'tcp',
        src       => '$EXTERNAL_NET',
        src_port  => 'any',
        direction => '->',
        dst       => '$HOME_NET',
        dst_port  => 'any',
        opts      => [
            ['msg'     => '"perl 6 download detected\; may the world rejoice!"'],
            ['depth'   => 150],
            ['offset'  => 0],
            ['content' => 'perl-6.0.0'],
            ['nocase'],
        ],
    };

=back

=cut

sub new {
    my ( $class, $data ) = @_;

    my $self = {
    };

    bless $self, $class;
    $self->_init($data);
}

=for comment

The _init method is called by the new method, to figure out what sort of data was passed to C<new()>.  If necessary, it calls $self->parse(), individual element accessor methods, or simply returns $self.

=cut

sub _init {
    my ( $self, $data ) = @_;

    # were we passed a hashref? (formatted rule in hashref form)
    if ( ref($data) eq "HASH" ) {
        # loop through the bits and set the values
        while ( my ( $method, $val ) = each %$data ) {
            $self->$method($val);
        }
    } elsif ( defined($data) ) {
        # otherwise, interpret this as a plain text rule.
        $self->parse($data);
    }
    # nothing
    return $self;
}

=head2 parse( $rule_string )

The parse method is what interprets a plain text rule, and populates the rule object.  Beacuse this module does not support the UNIX style line-continuations (backslash at the end of a line) the rule must be all on one line, otherwise the parse will fail in unpredictably interesting and confusing ways.  The parse method tries to interpret the rule from left to right, calling the individual accessor methods for each rule element.  This will overwrite the contents of the object (if any), so if you want to parse multiple rules at once, you will need multiple objects.

  $rule->parse($rule_string);

=cut

sub parse {
    my ( $self, $rule ) = @_;

    # nuke extra whitespace pre/post rule
    $rule =~ s/^\s+//;
    $rule =~ s/\s+$//;

    # Rules are distributed without being enabled
    if ($rule =~ /^#/) {
        $rule =~ s/^#+\s*//g;
        $self->state(0);
    }
    else {
        $self->state(1);
    }

    # 20090823 RGH: m/\s+/ instead of m/ /; bug reported by Leon Ward
    my @values = split(m/\s+/, $rule, scalar @RULE_ELEMENTS);    # no critic

    # Support for 'Decoder and Preprocessor Rules'
    if ($values[1] eq '(') {
        $self->{preprocessed} = 1;
        $self->action($values[0]);
        shift @values;
        $self->opts(join(' ', @values));
    }
    # Regular rules
    else {
        for my $i ( 0 .. $#values ) {
            my $meth = $RULE_ELEMENTS[$i];
            $self->$meth( $values[$i] );
        }
    }
}

=head2 state

The state of the rule: active (1) or commented (0)

=cut

sub state {
    my ($self, $state) = @_;

    if (defined $state) {
        $self->{state} = $state;
    }
    if (!defined $self->{state}) {
        return 1;
    }
    return $self->{state};
}

=head2 METHODS FOR ACCESSING RULE ELEMENTS

You can access the core parts of a rule (action, protocol, source IP, etc) with the method of their name.  These are read/write L<Class::Accessor> accessors.  If you want to read the value, don't pass an argument.  If you want to set the value, pass in the new value.  In either case it returns the current value, or undef if the value has not been set yet.

=for comment

Need to figure out "truth" again in perl sense, do I simply "return;" or "return undef" if the value doesn't exist?  For Parse::Snort::Strict, I need to have two things: 1) make it known to the user that the rule failed to parse, 2) which (may?) be a different meaning than the rule element being empty/undefined.

=over 4

=item action 

The rule action.  Generally one of the following: C<alert>, C<pass>, C<drop>, C<sdrop>, or C<log>.

=item proto

The protocol of the rule.  Generally one of the following: C<tcp>, C<udp>, C<ip>, or C<icmp>.

=item src

The source IP address for the rule.  Generally a dotted decimal IP address, Snort $HOME_NET variable, or CIDR block notation.

=item src_port

The source port for the rule.  Generally a static port, or a contigious range of ports.

=item direction

The direction of the rule.  One of the following: C<< -> >> C<< <> >> or C<< <- >>.

=item dst

The destination IP address for the rule.  Same format as C<src>

=item dst_port

The destination port for the rule.  Same format as C<src>

=item opts ( $opts_array_ref )

=item opts ( $opts_string )

The opts method can be used to read existing options of a parsed rule, or set them.  The method takes two forms of arguments, either an Array of Arrays, or a rule string.

=over 4

=item $opts_array_ref

  $opts_array_ref = [
       [ 'msg' => '"perl 6 download detected\; may the world rejoice!"' ],
       [ 'depth' => 150 ],
       [ 'offset' => 0 ],
       [ 'content' => 'perl-6.0.0' ],
       [ 'nocase' ],
  ]

=item $opts_string

  $opts_string='(msg:"perl 6 download detected\; may the world rejoice!";depth:150; offset:0; content:"perl-6.0.0"; nocase;)';

The parenthesis surround the series of C<key:value;> pairs are optional.

=back

=cut

sub opts {
    my ( $self, $args ) = @_;

    if ($args) {

        # setting
        if ( ref($args) eq "ARRAY" ) {

            # list interface:
            # ([depth => 50], [offset => 0], [content => "perl6"], ["nocase"])
            $self->set( 'opts', $args );
        } else {

            # string interface
            # 'depth:50; offset:0; content:"perl\;6"; nocase;'
            if ( $args =~ m/^\(/ ) {
              # remove opts parens if they exist
                $args =~ s/^\((.+)\)$/$1/;
            }

            # When I first wrote this regex I thought it was slick.
            # I still think that, but 2y after doing it the first time
            # it just hurt to look at.  So, /x modifier we go!
            my @set = map { [ split( m/\s*:\s*/, $_, 2 ) ] } $args =~ m/
                \s*         # ignore preceeding whitespace
                (           # begin capturing
                 (?:        # grab characters we want
                     \\.    # skip over escapes
                     |
                     [^;]   # or anything but a ;
                 )+?        # ? greedyness hack lets the \s* actually match
                )           # end capturing
                \s*         # ignore whitespace between value and ; or end of line
                (?:         # stop anchor at ...
                  ;         # semicolon
                  |         # or
                  $         # end of line
                )
                \s*/gx;
            $self->set( 'opts', @set );
        }
    } else {
        # getting
        return $self->get('opts');
    }
}

sub _single_opt_accessor {
    my $opt = shift;
    return sub {
        my ( $self, $val ) = @_;

        # find the (hopefully) pre-existing option in the opts AoA
        my $element;

        if ( defined $self->get('opts') ) {
            $element = first { $_->[0] eq $opt } @{ $self->get('opts') };
        }

        if ( ref($element) ) {

            # preexisting
            if ($val) { $element->[1] = $val; }
            else { return $element->[1]; }
        } else {

            # doesn't exist
            if ($val) {

                # setting
                if ( scalar $self->get('opts') ) {

                    # other opts exist, tack it on the end
                    $self->set(
                        'opts',
                        @{ $self->get('opts') },
                        [ $opt, $val ]
                    );
                } else {

                    # blank slate, create the AoA
                    $self->set( 'opts', [ [ $opt, $val ] ] );
                }
            } else {

                # getting
                return;
            }
        }
      }
}

# helper accessors that poke around inside rule options

*sid       = _single_opt_accessor('sid');
*rev       = _single_opt_accessor('rev');
*msg       = _single_opt_accessor('msg');
*classtype = _single_opt_accessor('classtype');
*gid       = _single_opt_accessor('gid');
*metadata  = _single_opt_accessor('metadata');
*priority  = _single_opt_accessor('priority');
*flow      = _single_opt_accessor('flow');

=back

=head2 HELPER METHODS FOR VARIOUS OPTIONS

=over 4

=item sid

=item rev

=item msg

=item classtype

=item gid

=item metadata

=item flow

=item priority

The these methods allow direct access to the rule option of the same name

  my $sid = $rule_obj->sid(); # reads the sid of the rule
  $rule_obj->sid($sid); # sets the sid of the rule
  ... etc ...

=item references

The C<references> method permits read-only access to the C<reference:> options in the rule.  This is in the form of an array of arrays, with each reference in the format

  [ 'reference_type' => 'reference_value' ]

To modify references, use the C<opts> method to grab all the rule options, modify it to your needs, and use the C<opts> method to save your changes back to the rule object.


  $references = $rule->references(); # just the references
  $no_references = grep { $_->[0] ne "reference" } @{ $rule->opts() }; # everything but the references

=cut 

sub references {
    my ($self) = shift;
    return [
        map { [split(m/,/, $_->[1], 2)] }
        grep { $_->[0] eq "reference" } @{ $self->get('opts') }
    ];
}

=item as_string

The C<as_string> method returns a string that matches the normal Snort rule form of the object.  This is what you want to use to write a rule to an output file that will be read by Snort.

=back

=cut

sub as_string {
    my $self = shift;
    my $ret;
    my @missing;

    # we may be incomplete
    @missing = grep { $_ } map { exists( $self->{$_} ) ? undef : $_ } @RULE_ELEMENTS_REQUIRED;

    # stitch together the required bits
    if (!scalar @missing) {
        $ret .= sprintf("%s %s %s %s %s %s %s",
            @$self{@RULE_ELEMENTS_REQUIRED});
    }

    # tack on opts if they exist
    if (defined $self->get('opts')) {
        $ret .= sprintf(
            " (%s)",
            join(" ",
                map { defined($_->[1]) ? "$_->[0]:$_->[1];" : "$_->[0];" }
                    @{ $self->get('opts') })
        );
    }

    return undef if @missing && !$self->{preprocessed};
    return $self->state ? $ret : "# $ret";
}

=head1 AUTHOR

Richard G Harman Jr, C<< <perl-cpan at richardharman.com> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-parse-snort at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Parse-Snort>.
I will be notified, and then you' ll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Parse::Snort

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Parse-Snort>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Parse-Snort>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Parse-Snort>

=item * Search CPAN

L<http://search.cpan.org/dist/Parse-Snort>

=back

=head1 DEPENDENCIES

L<Class::Accessor>, L<List::Util>, L<Sub::Util>, L<Carp>

=head1 ACKNOWLEDGEMENTS

MagNET #perl for putting up with me :)

=head1 COPYRIGHT & LICENSE

Copyright 2007 Richard Harman, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

!!'mtfnpy!!';
